### DISCUSSION TITLE
Episode 188: Parallel std::merge

### DISCUSSION THREAD

**Discussion by codereport** [codereport/adsp2#87] [Episode]
Episode Link: https://adspthepodcast.com/2024/06/28/Episode-188.html

Feel free to write comments or ask questions 🙂

**Comment by mermerico** [comment]
Here's how I'm thinking about the problem: in the sequential algorithm there is an outer loop over the output elements and an inner loop over the input lists (that chooses the min or max element). You can obviously parallelize the inner loop using a classic parallel reduction. But I think you can also parallelize the outer loop. Let's say we are trying to sort three input lists L1, L2, and L3. Our elements range in value between 0 and 100. We could divide input lists into chunks at certain boundary values, like 33 and 66. We now have 9 input lists: L1_0 (with values between 0 and 32), L1_33 with values between 33 and 65, L1_66, L2_0, etc. Now we can merge L1_0, L2_0 and L3_0 together while in parallel merging L1_33, L2_33, L3_33 and also in parallel merging the LX_66s. Since the input lists are sorted, finding a good set of boundaries in L1 is quick, as is finding the corresponding boundary locations in L2 and L3. Once you have merged lists L_0, L_33 and L_66, you can just concatenate them together.

**Comment by roibarkan** [comment]
The safari team/lab at ETH Zurich has some great academic knowledge and material on parallel algorithms with GPUs and other platforms.
Here's a lecture specifically about merge: https://youtu.be/14Nd0MWY1Vg

**Reply by mermerico** [reply]
This is a really great resource, thanks for sharing! This was the algorithm I was trying to describe!
